src/metabase/driver/clickhouse_qp.clj

```diff
@@ -0,0 +1,429 @@
+   :else nil])
+
+(defmethod sql.qp/->honeysql [:clickhouse :sum-where]
+  [driver [_ field pred]]
+  [:sum [:case (sql.qp/->honeysql driver pred) (sql.qp/->honeysql driver field)
+         :else 0]])
+
+(defmethod sql.qp/add-interval-honeysql-form :clickhouse
+  [_ dt amount unit]
+  (h2x/+ (h2x/->timestamp dt)
+         [:raw (format "INTERVAL %d %s" (int amount) (name unit))]))
+
+;; The following lines make sure we call lowerUTF8 instead of lower
+(defn- ch-like-clause
+  [driver field value options]
+  (if (get options :case-sensitive true)
+    [:like field (sql.qp/->honeysql driver value)]
+    [:like [:'lowerUTF8 field]
+     (sql.qp/->honeysql driver (update value 1 metabase.util/lower-case-en))]))
+
+(s/defn ^:private update-string-value :- mbql.s/value
+  [value :- (s/constrained mbql.s/value #(string? (second %)) ":string value") f]
+  (update value 1 f))
+
+(defmethod sql.qp/->honeysql [:clickhouse :contains]
+  [driver [_ field value options]]
+  (ch-like-clause driver
+                  (sql.qp/->honeysql driver field)
+                  (update-string-value value #(str \% % \%))
+                  options))
+
+(defn- clickhouse-string-fn
+  [fn-name field value options]
+  (let [field (sql.qp/->honeysql :clickhouse field)
+        value (sql.qp/->honeysql :clickhouse value)]
+    (if (get options :case-sensitive true)
+      [fn-name field value]
+      [fn-name [:'lowerUTF8 field] (metabase.util/lower-case-en value)])))
+
+(defmethod sql.qp/->honeysql [:clickhouse :starts-with]
+  [_ [_ field value options]]
+  (clickhouse-string-fn :'startsWith field value options))
+
+(defmethod sql.qp/->honeysql [:clickhouse :ends-with]
+  [_ [_ field value options]]
+  (clickhouse-string-fn :'endsWith field value options))
+
+;; We do not have Time data types, so we cheat a little bit
+(defmethod sql.qp/cast-temporal-string [:clickhouse :Coercion/ISO8601->Time]
+  [_driver _special_type expr]
+  (h2x/->timestamp [:'parseDateTimeBestEffort [:'concat "1970-01-01T" expr]]))
+
+(defmethod sql.qp/cast-temporal-byte [:clickhouse :Coercion/ISO8601->Time]
+  [_driver _special_type expr]
+  (h2x/->timestamp expr))
+
+(defmethod sql-jdbc.execute/read-column-thunk [:clickhouse Types/TINYINT]
+  [_ ^ResultSet rs ^ResultSetMetaData _ ^Integer i]
+  (fn []
+    (.getByte rs i)))
+
+(defmethod sql-jdbc.execute/read-column-thunk [:clickhouse Types/SMALLINT]
+  [_ ^ResultSet rs ^ResultSetMetaData _ ^Integer i]
+  (fn []
+    (.getShort rs i)))
+
+;; This is for tests only - some of them expect nil values
+;; getInt/getLong return 0 in case of a NULL value in the result set
+;; the only way to check if it was actually NULL - call ResultSet.wasNull afterwards
+(defn- with-null-check
+  [^ResultSet rs get-value-fn]
+  (let [value (get-value-fn)]
+    (if (.wasNull rs) nil value)))
+
+(defmethod sql-jdbc.execute/read-column-thunk [:clickhouse Types/BIGINT]
+  [_ ^ResultSet rs ^ResultSetMetaData _ ^Integer i]
+  (fn []
+    (with-null-check rs #(.getLong rs i))))
```

Comment:
I will check :+1: This piece is an ancient one.

Reply:
Wasn't `read-column` deprecated and removed since IIRC 0.46? 
https://github.com/metabase/metabase/blob/v0.45.4.3/src/metabase/driver/sql_jdbc/execute/old_impl.clj#L27-L42
So we have to use the thunk one now.

Reply:
Sure. I just meant changing `with-null-check` to receive the result value instead of a function pulling the result value. On one hand that's simpler and possibly faster, on the other hand makes it a little bit harder to see that the null check immediately follows pulling the result value.